### PR TITLE
Web3 pending transaction count

### PR DESCRIPTION
**Pull request by hkalodner**
I pulled this back to be a draft since the arb-test-aggregator is experiencing bugs due to an issue with the SimulatedBackend

- Add support for querying the pending transaction count
- Fix `GetTransactionReceipt` correctness
- Add `arb-test-aggregator` which is a self-contained blockchain and tx-aggregator process allowing for easier testing
- Fix JSON test vector creation for empty tuples

### CODE REVIEW

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 8
- Added 61
           

Complexity increasing per file
==============================
- packages/arb-tx-aggregator/rpc/launch.go  5
- packages/arb-tx-aggregator/web3/eth.go  3
- packages/arb-tx-aggregator/testvector/testvector.go  14
- packages/arb-tx-aggregator/cmd/arb-test-aggregator/arb-test-aggregator.go  21
- packages/arb-tx-aggregator/aggregator/aggregator.go  1
- packages/arb-tx-aggregator/cmd/arb-check-vector/arb-check-vector.go  2
- packages/arb-util/value/tuple.go  1
         

Complexity decreasing per file
==============================
+ packages/arb-validator-core/ethbridge/chain.go  -3
+ packages/arb-tx-aggregator/machineobserver/observermanager.go  -5
+ packages/arb-tx-aggregator/arbostest/testlogs_test.go  -12
         

Clones added
============
- packages/arb-validator-core/ethbridge/globalInbox.go  2
- packages/arb-tx-aggregator/cmd/arb-test-aggregator/arb-test-aggregator.go  1
- tests/fibgo/connection_test.go  1
         

Clones removed
==============
+ packages/arb-tx-aggregator/machineobserver/observermanager.go  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=5984973&bid=19625795)

File: packages/arb-tx-aggregator/web3/eth.go [hunk 1]
```diff
@@ -117,31 +127,56 @@ func (s *Server) GetCode(r *http.Request, args *AccountInfoArgs, reply *string)
 	return nil
 }
 
-func (s *Server) blockNum(ctx context.Context, block *rpc.BlockNumber) (uint64, error) {
-	if *block == rpc.LatestBlockNumber {
-		return s.srv.GetBlockCount(ctx)
+func (s *Server) blockNumRestrictLatest(ctx context.Context, block *rpc.BlockNumber) (uint64, error) {
```
Comment (edfelten) [inline packages/arb-tx-aggregator/web3/eth.go:130]: Add a comment explaining what this does and how it differs from `BlockNum`.

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 8
- Added 61
           

Complexity increasing per file
==============================
- packages/arb-tx-aggregator/cmd/arb-check-vector/arb-check-vector.go  2
- packages/arb-util/value/tuple.go  1
- packages/arb-tx-aggregator/txdb/txdb.go  1
- packages/arb-tx-aggregator/rpc/launch.go  4
- packages/arb-tx-aggregator/testvector/testvector.go  14
- packages/arb-tx-aggregator/aggregator/aggregator.go  1
- packages/arb-tx-aggregator/cmd/arb-test-aggregator/arb-test-aggregator.go  21
- packages/arb-tx-aggregator/web3/eth.go  3
         

Complexity decreasing per file
==============================
+ packages/arb-tx-aggregator/arbostest/testlogs_test.go  -12
+ packages/arb-tx-aggregator/machineobserver/observermanager.go  -4
+ packages/arb-validator-core/ethbridge/chain.go  -3
         

Clones added
============
- packages/arb-tx-aggregator/cmd/arb-test-aggregator/arb-test-aggregator.go  1
- packages/arb-validator-core/ethbridge/globalInbox.go  2
- tests/fibgo/connection_test.go  1
         

Clones removed
==============
+ packages/arb-tx-aggregator/machineobserver/observermanager.go  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=5984973&bid=19625795)

**Comment by hkalodner** [comment]
This PR was superseded

**Pull request closed without merging**